File: go/src/solutions/two_sum_test.go
// LeetCode 1. Two Sum in Go

package solutions

import (
	"reflect"
	"testing"
)

func twoSum(nums []int, target int) []int {
	indices := make(map[int]int)
	for i, num := range nums {
		if indice, ok := indices[target-num]; !ok {
			indices[num] = i
		} else {
			return []int{indice, i}
		}
	}
	return nil
}

func TestTwoSum(t *testing.T) {
	var tests = []struct {
		nums   []int
		target int
		ans    []int
	}{
		{[]int{2, 7, 11, 15}, 9, []int{0, 1}},
		{[]int{3, 2, 4}, 6, []int{1, 2}},
		{[]int{3, 3}, 6, []int{0, 1}},
	}

	for _, test := range tests {
		if !reflect.DeepEqual(twoSum(test.nums, test.target), test.ans) {
			t.Errorf("twoSum(%v, %d) != %v", test.nums, test.target, test.ans)
		}
	}
}

File: python/src/solutions/two_sum_test.py
# LeetCode 1. Two Sum

import unittest
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        indices = dict()
        for i in range(0, len(nums)):
            if (target - nums[i]) in indices:
                return [indices[target - nums[i]], i]
            else:
                indices[nums[i]] = i
        return []


class TestTwoSum(unittest.TestCase):

    def testExample1(self):
        nums = [2, 7, 11, 15]
        target = 9
        ans = [0, 1]
        self.assertEqual(Solution().twoSum(nums, target), ans)

    def testExample2(self):
        nums = [3, 2, 4]
        target = 6
        ans = [1, 2]
        self.assertEqual(Solution().twoSum(nums, target), ans)

    def testExample3(self):
        nums = [3, 3]
        target = 6
        ans = [0, 1]
        self.assertEqual(Solution().twoSum(nums, target), ans)
